### ISSUE TITLE
Gradients be dropped in tf.train.SyncReplicaOptimizer

### ISSUE THREAD

**Issue by hongjic** [tensorflow/tensorflow#11660]
### Problem Description
tf.train.SyncReplicaOptimizer is very helpful for backup workers, but the gradients computed by slow workers are just dropped. Below is the documentation I see

> Once the gradients have been computed, push them into gradient_queue only if local_step equals global_step, otherwise the gradients are just dropped

The problem is, I want to use backup workers in online training, it means every sample can be just consumed by TensorFlow once and I don't want some samples to be just ignored, is it possible to make backup servers behave like the one in Hadoop which is guaranteed to utilize all data partitions?
Or can the workers know whether its gradient is dropped or not by the aggregator?
I think this feature will be useful.

**Comment by asimshankar** [comment]
I am perhaps misunderstanding, but isn't the point of the backup workers to provide a tradeoff between waiting for replicas and consuming all data (so, the step proceeds when K of the N workers have updates, where the number of backup workers is (N-K)). If you want data from all the workers to apply, then increase the number of workers and set backup workers to zero (i.e., set [`replicas_to_aggregate` and leave `total_num_replicas` as None](https://www.tensorflow.org/api_docs/python/tf/train/SyncReplicasOptimizer#__init__).)

Or am I missing something?

**Comment by hongjic** [comment]
@asimshankar you are right, but I hope it can consume all data but at the same time use backup servers to enhance training speed of each step.
It can be as simple as let the worker recosume the data that was dropped.

**Comment by asimshankar** [comment]
There are no input guarantees provided in distributed TensorFlow at this time (i.e., no guarantees that every example has been seen exactly once in an epoch). It is something we're looking into, but no concrete timeline. 

Doing that and then being able to ensure that samples trained on by the backup workers are picked up in future steps by other workers is something that will require a detailed design and proposal. I'm tempted to close this issue since it isn't a simple feature request.

If you'd like to contribute this, I would encourage you to think through a possible design and share detailed design thoughts here. Thanks!

**Comment by hongjic** [comment]
Thank you @asimshankar  I would love to spend some time on it.
